api/s3/s3.go

```diff
@@ -0,0 +1,289 @@
+}
+
+// GetBucketLocation respondes to an S3 GetBucketLocation request. This request
+// only contains the AWS region name as XML.
+// https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetBucketLocation.html
+func GetBucketLocation(c *gin.Context) {
+	log.Debug("S3 GetBucketLocation request")
+
+	// Gin doesn't write the xml header when using c.XML, so we add it manually
+	_, err := c.Writer.Write([]byte(xml.Header))
+	if err != nil {
+		log.Errorf("Failed writing XML header: %v", err)
+		c.AbortWithStatus(http.StatusInternalServerError)
+
+		return
+	}
+	c.XML(http.StatusAccepted, LocationConstraint{
+		XMLns:    "http://s3.amazonaws.com/doc/2006-03-01/",
+		Location: "us-west-2",
+	})
+}
+
+// ListBuckets respondes to an S3 ListBuckets request. This request returns the
+// available S3 buckets. We use this to list accessible datasets.
+// https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListBuckets.html
+func ListBuckets(c *gin.Context) {
+	log.Debug("S3 ListBuckets request")
+
+	// Gin doesn't write the xml header when using c.XML, so we add it manually
+	_, err := c.Writer.Write([]byte(xml.Header))
+	if err != nil {
+		log.Errorf("Failed writing XML header: %v", err)
+		c.AbortWithStatus(http.StatusInternalServerError)
+
+		return
+	}
+
+	buckets := []Bucket{}
+	for _, dataset := range middleware.GetDatasets(c) {
+		datasetInfo, err := database.GetDatasetInfo(dataset)
+		if err != nil {
+			log.Errorf("Failed to get dataset information: %v", err)
+			c.AbortWithStatus(http.StatusInternalServerError)
+		}
+		// TODO: Add real creation date
+		buckets = append(buckets, Bucket{Name: datasetInfo.DatasetID,
+			CreationDate: datasetInfo.CreatedAt})
+	}
+
+	c.XML(http.StatusAccepted, ListAllMyBucketsResult{
+		Buckets: buckets,
+		Owner:   Owner{DisplayName: "", ID: ""},
+	})
+}
+
+// ListObjects respondes to an S3 ListObjects request. This request
+// lists the contents of an S3 bucket. We use this to return dataset content.
+// https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjects.html
+func ListObjects(c *gin.Context) {
+	log.Debug("S3 ListObjects request")
+
+	dataset := c.Param("dataset")
+
+	allowed := false
+	for _, known := range middleware.GetDatasets(c) {
+		if dataset == known {
+			allowed = true
+
+			break
+		}
+	}
+	if !allowed {
+		c.AbortWithStatus(http.StatusNotFound)
+
+		return
+	}
+
+	files, err := database.GetFiles(dataset)
+	if err != nil {
+		log.Errorf("Failed getting dataset files: %v", err)
+		c.AbortWithStatus(http.StatusInternalServerError)
+
+		return
+	}
+
+	// We return the full upload path, as file key
+	objects := []Object{}
+	for _, file := range files {
+		lastModified, err := time.Parse(time.RFC3339, file.LastModified)
+		if err != nil {
+			log.Errorf("failed to parse last modified time: %v", err)
+			c.AbortWithStatus(http.StatusInternalServerError)
+
+			return
+		}
+		objects = append(objects, Object{
+			Key:          strings.TrimSuffix(file.FilePath, ".c4gh"),
```

Comment:
```suggestion
			Key:          file.FileID,
```

This needs to return the file stable ID

Reply:
What do you mean? That would completely break the functionality!

Reply:
this query:
```
s3cmd -c download.ini ls s3://EGAD74900000101/
```

should return this structure:
```
2023-04-26 14:42   14M  s3://EGAD74900000101/EGAF74900000001
2023-04-26 14:42    8M  s3://EGAD74900000101/EGAF74900000002
```

Reply:
No, it should not. Please read the PR description.

Reply:
I know what is says, but this is breaking established semantics from the HTTP API.